crates/service-protocol/src/discovery.rs

```diff
@@ -268,6 +273,21 @@ impl ServiceDiscovery {
             }
         };
 
+        match (protocol_type, endpoint) {
+            // all endpoints support request response
+            (ProtocolType::RequestResponse, _) => {}
+            // http2 supports bidi
+            (ProtocolType::BidiStream, Endpoint::Http(_, hyper::Version::HTTP_2)) => {}
+            // alpn client is not guaranteed to support bidi
+            (ProtocolType::BidiStream, Endpoint::Http(_, _)) => {
+                return Err(DiscoveryError::BidirectionalRequiresHTTP2);
+            }
```

Comment:
if they set --use-http1.1, it would be probably incorrect to then persist a value of http2-prior-knowledge. some endpoints may support http2 upgrade via alpn but not prior knowledge, who knows. and if we instead persist http1.1, but allow bidi because we observed a http2 upgrade, we dont know this will *continue* to be true as the user has explicitly requested that we do not assume that the endpoint will always serve http2

Reply:
Ah ok, so if the user says `--use-http1.1`, then it should always use the ALPN client to figure this out at runtime.